### ISSUE TITLE
Can't download files uploaded with megaput

### ISSUE THREAD

**Issue by lambdacerro** [megous/megatools#242]
Hello,

I've been trying to use megaput to upload files to mega, to later download them via the mega website, however im not having much luck.

I'm using the megatools in the Ubuntu repository, also tried with the compiled 1.9.98 directly from the website (https://megatools.megous.com/)

The error is always the same, the upload ends just fine, however when downloading the file from the mega website it fails to decrypt the file.

Happens on any kind of file, i've tried with 200MB, 800MB and 3GB files, all with the same result, looks like on smaller files it doesn't happen that often, but it does happen also.

megadl suffers from the same issue, fails with the MAC mismatch error

Basically, to put in short.

megadl can't download files uploaded with megaput due to a MAC mismatch, neither can the MEGA website download them.

**Comment by lambdacerro** [comment]
Such a weird thing, i've been doing more tests and something is weird.

The binaries on machine A, gives the mac error when downloading.
The same EXACT binaries on machine B, works fine.

Tried the same binaries on other machines, same hardware, without issues.

Looks like the issue was with the original machine, however, i have no clue of why it happened at first, i've formatted the machine and that solved the issue, I guess this will end up being a mistery.

I'll close the issue as probably isn't related to megatools itself, probably some library acting weird.

**Comment by megous** [comment]
Thanks for the info. I started testing it and it seemed to work fine for me. So perhaps it could have been some corruption in some binary, or whatever.

**Comment by lambdacerro** [comment]
Hello, I'm reopening this because I think that there's something wrong in the code and happens from time to time.

Theres some files that uploads and downloads just fine on the mega website, however, some artifacts happens when decrypting the file attributes, for example, in some files where padding is needed for the attributes, the attributes decrypts fine, however, the padding is wrong, of the attrs is 10 chars in length, the rest 6 bytes should be null values, however, theres a NULL, a ascii char just after it and the rest is null bytes, as it should, example.

MEGA{"n": "filename.txt"}NULLmNULLNULLNULLNULLNULL

I use a php library made by me to download mega files, it always worked just fine with files uploaded on the mega website, but started having issues with some files uploaded by megatools.

The issue was related with the decrypting of the attributes, because the library unpads the attrs by removing all null characters starting from the end of the string, however, when it finds the random ascii char it stops there and gives out this

MEGA{"n": "filename.txt"}NULLm

Thus making the json parser fail because of that m, the mega website pads the attrs correctly, everything after the } is NULL bytes

Something is corrupting the encrypt of the attributes, and I think, that the same issue is what makes the MAC verification fail, the mac is probably computed correctly but fails to decrypt for whatever reason.

The issue doesn't happen always but in some files, which is weird, i've took a look at the C source of mega_session_put and the function that encrypts the attributes and couldn't find anything wrong, will continue to do testing and will update with any findings.

**Comment by megous** [comment]
You're right, there's a bug in b64_aes128_cbc_encrypt_str. data is a zeroed buffer, but the source string is copied with some uninitialized memory past the end of it.

**Comment by megous** [comment]
memcpy must use the original len before it is aligned to 16 bytes

**Comment by megous** [comment]
I guess this only matters if you strip null bytes from the end, and not search for the null byte from the beginning. But obviously this needs to be fixed.

**Comment by lambdacerro** [comment]
Yep, just found this, I made b64_aes128_cbc_encrypt_str save all the strings to file and the padding got wrong bytes sometimes, for example, in a string 10177 bytes in length, 15 bytes are needed to pad the string, and its padded this way.

}]}NULLNULLNULLNULLNULLNULLNULLá'NULLNULLNULLNULLNULLNULL

**Comment by megous** [comment]
Fix is in the stable branch.

**Comment by megous** [comment]
Thanks for the report. :)

**Comment by lambdacerro** [comment]
No problem, will close this for now, lets see if this is what was making the mega website fail to download the files, but something tells me is not, since the megadl itself fails with the MAC Mismatch error, in any case i'll reopen the issue with more information.

**Comment by lambdacerro** [comment]
Yep, same thing is still happening, some files uploads fine, other uploads correclty (the file is actually decrypted just fine) however the MAC check fails, I dont know if its because the file itself uploaded with some bytes wrong or the mac itself is not computed correctly.

**Comment by lambdacerro** [comment]
Doing some testing, will see if the issue is random or happens with files of a determined size

**Comment by megous** [comment]
The MAC issue has nothing to do with attributes decoding/encoding.

**Comment by lambdacerro** [comment]
Yep, I know, it's a different issue.

**Comment by lambdacerro** [comment]
I made some progress, I uploaded a file that downloaded gives out the mac mismatch, then i md5'd both files (the original and the downloaded with the mac mismatch) and both gives different checksums, gonna diff the files to see what differences are there.

**Comment by lambdacerro** [comment]
The files are basically the same except for two blocks with different bytes each one, weird thing, so probably the error is @ the uploading of the data

**Comment by lambdacerro** [comment]
I'm lost in this one, can't see anything wrong in the code that could cause the file to get uploaded wrong.

I've just uploaded the exact same original file and worked just fine, it just happens randomly and can't figure it out why it happens.

**Comment by megous** [comment]
I'll make some stress test (upload+download) with files of various sizes. Do you remember if the blocks were located at the end of the file, by any chance?

**Comment by megous** [comment]
Upload/download code paths were re-factored. I've also added stress tests, and can't hit any issues.
